proving.go

```diff
@@ -94,37 +95,40 @@ func calcSubtreeProof(cache *TreeCache, hash HashFunc, leavesToProve []uint64, s
 		return nil, errors.New("while preparing to traverse subtree: " + err.Error())
 	}
 
-	additionalProof, _, err := traverseSubtree(reader, width, hash, relativeLeavesToProve)
+	_, additionalProof, err := traverseSubtree(reader, width, hash, relativeLeavesToProve, nil)
 	if err != nil {
 		return nil, errors.New("while traversing subtree: " + err.Error())
 	}
 
 	return additionalProof, err
 }
 
-func traverseSubtree(leafReader NodeReader, width uint64, hash HashFunc,
-	leavesToProve []uint64) ([][]byte, []byte, error) {
+func traverseSubtree(leafReader NodeReader, width uint64, hash HashFunc, leavesToProve []uint64, paddingValue []byte) (
+	[]byte, [][]byte, error) {
 
-	t := NewProvingTree(hash, leavesToProve)
+	t := NewTreeBuilder(hash).
+		WithLeavesToProve(leavesToProve).
+		WithMinHeight(rootHeightFromWidth(width)). // This ensures the correct size tree, even if padding is needed.
+		Build()
 	for i := uint64(0); i < width; i++ {
 		leaf, err := leafReader.ReadNext()
-		if err != nil {
+		if err == io.EOF {
+			// Add external padding if provided.
+			if paddingValue == nil {
+				break
+			}
+			leaf = paddingValue
+			paddingValue = nil
```

Comment:
input vars are passed by value, so this doesn't affect the outer scope. It would if I'd have changed the contents of the underlying array, but nullifying the slice is fine.

If you don't like if for "softer" reasons (i.e. it's just weird / bad practice) then I can add a `bool` called `usedPadding` and use that to flag that I should use it again. I like the way it works right now, but if it bothers you - lmk.

Reply:
Yes, there is no practical reason behind this comment, only a practice issue. Again, up to you to decide.